nomos-services/cryptarchia-consensus/src/leadership.rs

```diff
@@ -27,87 +24,83 @@ pub struct LeaderConfig {
-            notes: HashMap::from([(header_id, header_notes)]),
+            notes,
             nf_sk,
             config,
         }
     }
 
-    // Signal that the chain extended with a new header, possibly evolving a leader
-    // notes in the process FIXME: this implementation does not delete old coins
-    // and will attempt to re-use a note in different forks,        we should
-    // use the orphan proofs mechanism to handle this.
-    pub fn follow_chain(&mut self, parent_id: HeaderId, id: HeaderId, to_evolve: Nullifier) {
-        if let Some(notes) = self.notes.get(&parent_id) {
-            let notes = notes
-                .iter()
-                .map(|note| {
-                    let note_cm = note.commit(self.nf_sk.commit());
-                    if Nullifier::new(self.nf_sk, note_cm) == to_evolve {
-                        evolve(note, self.nf_sk)
-                    } else {
-                        *note
-                    }
-                })
-                .collect();
-            self.notes.insert(id, notes);
-        }
-    }
-
     pub async fn build_proof_for(
         &self,
-        note_tree: &NoteTree,
+        aged_tree: &NoteTree,
+        latest_tree: &NoteTree,
         epoch_state: &EpochState,
         slot: Slot,
-        parent: HeaderId,
     ) -> Option<Risc0LeaderProof> {
-        let notes = self.notes.get(&parent)?;
-        for note in notes {
+        fn find(tree: &NoteTree, note_commit: &NoteCommitment) -> Option<usize> {
+            tree.commitments().iter().position(|cm| cm == note_commit)
+        }
+
+        for note in &self.notes {
             let note_commit = note.commit(self.nf_sk.commit());
-            let Some(index) = note_tree
-                .commitments()
-                .iter()
-                .position(|cm| cm == &note_commit)
-            else {
-                continue;
-            };
 
-            let input_cm_path = note_tree
-                .witness(index)
-                .expect("Note was found in the tree");
-            let note = InputWitness {
-                note: *note,
-                nf_sk: self.nf_sk,
-                cm_path: input_cm_path,
-            };
+            match (
+                find(aged_tree, &note_commit),
+                find(latest_tree, &note_commit),
+            ) {
+                (Some(_), Some(_)) => {
+                    tracing::debug!("Found note in aged & lastest trees: {:?}", note);
+                }
+                (Some(_), _) => {
+                    tracing::debug!("Found note in aged tree only: {:?}", note);
+                    continue;
+                }
+                (_, Some(_)) => {
+                    tracing::debug!("Found note in latest tree only: {:?}", note);
+                    continue;
+                }
+                (None, None) => {
+                    tracing::debug!("Note not found in either tree: {:?}", note);
+                    continue;
+                }
+            }
```

Comment:
Nit: This may be a function. Although having it there is immediate info, it also pollutes the loop a bit. Up to you anyway.

Reply:
this is actually a debug leftover, I'll remove everything but the success case